dev-docs/yaml-design-guidance.md

```diff
@@ -0,0 +1,293 @@
+
+Note that these are not hard and fast rules, and if your config is really obviously correct then bu all means make the change and be prepared to deal with comments at the review stage.
+
+## Design patterns
+
+Use the following as a rule of thumb, also look at existing config for inspiration.
+The most important goal is usability, so do break the rules if it makes sense, but it's worth bringing the discussion to the team in such circumstances.  
+
+1. [Avoid empty config](#avoid-empty-config).
+2. [Do use `#[serde(deny_unknown_fields)]`](#do-use-serdedeny_unknown_fields).
+3. [Don't use `#[serde(flatten)]`](#dont-use-serdeflatten).
+4. [Use consistent terminology](#use-consistent-terminology).
+5. [Document your configuration options](#document-your-configuration-options).
+6. [Plan for the future](#plan-for-the-future).
+
+### Avoid empty config
+
+In Rust you can use `Option` to say that config is optional, however this can give a bad experience if the type is complex and all fields are optional.
+
+#### GOOD
+```rust
+#[serde(deny_unknown_fields)]
+struct Export {
+    url: Url // url is required
+}
+```
+```yaml
+export:
+    url: http://example.com
+```
+
+#### GOOD
+```rust
+enum ExportUrl {
+    Default,
+    Url(Url)
+}
+
+#[serde(deny_unknown_fields)]
+struct Export {
+    url: ExportUrl // Url is required but user may specify `default`
+}
+```
+```yaml
+export:
+    url: default
+```
+
+#### BAD
+```rust
+#[serde(deny_unknown_fields)]
+struct Export {
+    url: Optional<Url> // url is optional
+}
+```
+```yaml
+export: # The user is not aware that url was defaulted.
+```
+
+#### UGLY
+In the case where you genuinely have no config then it may be acceptable to have an `enabled: bool` flag.
+The reason to avoid this is that it creates a disconnect between the user and the thing that they are trying to do
+```rust
+#[serde(deny_unknown_fields)]
+struct Export {
+    enabled: bool,
+    url: Optional<Url> // url is optional
+}
+```
+```yaml
+export: 
+  enabled: true # The user is not aware that url was defaulted.
+```
+
+### Do use `#[serde(deny_unknown_fields)]`.
+Every container that takes part in config should be annotated with `#[serde(deny_unknown_fields)]`. If not the user can make mistakes on their config and they they won't get errors.
+
+#### GOOD
+```rust
+#[serde(deny_unknown_fields)]
+struct Export {
+    url: Url
+}
+```
+```yaml
+export: 
+  url: http://example.com
+  backup: http://example2.com # The user will receive an error for this
+```
+
+#### BAD
+```rust
+struct Export {
+    url: Url
+}
+```
+```yaml
+export: 
+  url: http://example.com
+  backup: http://example2.com # The user will NOT receive an error for this
+```
+
+### Don't use `#[serde(flatten)]`
+Serde flatten is tempting to use where you have identified common functionality, but creates a bad user experience as it is incompatible with `#[serde(deny_unknown_fields)]`. There isn't a great solution to this, but nesting config can sometimes help.
```

Comment:
that's an interesting point, I did not know about this. Should we remove the current uses of `serde(flatten)`?